### PR TITLE
feat: added open/close state of discussion sidebar in local storage

### PR DESCRIPTION
**Pull request by muhammadadeeltajamul**
- Discussion Sidebar will maintain its state i.e. if a user opens discussion sidebar in a unit and then navigates to course home. When he will open any unit, discussion sidebar will remain open. 
- User will atleast once view discussion sidebar.
- Sidebar information has been maintained in localstorage

Ticket: [INF-767](https://2u-internal.atlassian.net/browse/INF-767)

### CODE REVIEW

**Review by sundasnoreen12 — CHANGES_REQUESTED** [review]
PR LGTM. Please add test cases to store enable state of the sidebar in the local storage to improve the codeCoverage.

**Review by ayesha-waris — APPROVED** [review]
Test Coverage is failing please review.

File: src/courseware/course/Course.test.jsx [hunk 1]
```diff
@@ -43,6 +49,29 @@ describe('Course', () => {
...
   });
 
+  const setupDiscussionSidebar = async (storageValue = false) => {
+    localStorage.clear();
+    const testStore = await initializeTestStore({ provider: 'openedx' });
+    const state = testStore.getState();
+    const { courseware } = state;
+    const { courseId } = courseware;
+    const axiosMock = new MockAdapter(getAuthenticatedHttpClient());
+    axiosMock.onGet(`${getConfig().LMS_BASE_URL}/api/discussion/v1/courses/${courseId}`).reply(200, { provider: 'openedx' });
```
Comment (sundasnoreen12) [inline src/courseware/course/Course.test.jsx:58]: You can pass a static courseId as we passing in discussion MFE for api calling

File: src/courseware/course/Course.test.jsx [hunk 2]
```diff
@@ -43,6 +49,29 @@ describe('Course', () => {
     setItemSpy.mockRestore();
   });
 
+  const setupDiscussionSidebar = async (storageValue = false) => {
+    localStorage.clear();
+    const testStore = await initializeTestStore({ provider: 'openedx' });
+    const state = testStore.getState();
+    const { courseware } = state;
+    const { courseId } = courseware;
```
Comment (sundasnoreen12) [inline src/courseware/course/Course.test.jsx]: Use static courseId instead of getting from store as we are using in discussion MFE

File: src/courseware/course/Course.test.jsx [hunk 3]
```diff
@@ -43,6 +49,29 @@ describe('Course', () => {
...
+    const axiosMock = new MockAdapter(getAuthenticatedHttpClient());
+    axiosMock.onGet(`${getConfig().LMS_BASE_URL}/api/discussion/v1/courses/${courseId}`).reply(200, { provider: 'openedx' });
+    const topicsResponse = buildTopicsFromUnits(state.models.units);
+    axiosMock.onGet(`${getConfig().LMS_BASE_URL}/api/discussion/v2/course_topics/${courseId}`)
```
Comment (sundasnoreen12) [inline src/courseware/course/Course.test.jsx:60]: You can pass a static courseId as we passing in discussion MFE for api calling

File: src/courseware/course/Course.test.jsx [hunk 4]
```diff
@@ -186,6 +218,31 @@ describe('Course', () => {
     expect(screen.getByText(Object.values(models.sequences)[0].title)).toBeInTheDocument();
   });
 
+  [true, false, null].forEach(async (value) => (it(`discussion sidebar is ${value === false ? 'hidden' : 'shown'} when localstorage value is ${value}`, async () => {
+    await setupDiscussionSidebar(value);
+    const elementTestId = 'sidebar-DISCUSSIONS';
+    await waitFor(() => screen.queryByTestId(elementTestId));
+    const element = screen.queryByTestId(elementTestId);
```
Comment (sundasnoreen12) [inline src/courseware/course/Course.test.jsx]: use this instead of these lines to avoid duplication.
const element = await waitFor(() => screen.queryByTestId('sidebar-DISCUSSIONS'));

File: src/courseware/course/Course.test.jsx [hunk 5]
```diff
@@ -186,6 +218,31 @@ describe('Course', () => {
...
+    if (value === false) {
+      expect(element).toHaveClass('d-none');
+    } else {
+      expect(element).not.toHaveClass('d-none');
+    }
+  })));
+
+  [true, false].forEach(async (value) => (it(`Discussion sidebar storage value is ${!value} when sidebar is ${value ? 'closed' : 'open'}`, async () => {
```
Comment (sundasnoreen12) [inline src/courseware/course/Course.test.jsx]: use object instead of bool to pass the result attribute to avoid if else condition 
[{value: true, result: 'false'}, {value: false, result: 'true'}].forEach(async ({value, result})

File: src/courseware/course/Course.test.jsx [hunk 6]
```diff
@@ -186,6 +218,31 @@ describe('Course', () => {
...
+    await setupDiscussionSidebar(value);
+    await act(async () => {
+      const button = await screen.queryByRole('button', { name: /Show discussions tray/i });
+      button.click();
+    });
+    if (value === true) {
+      expect(localStorage.getItem('showDiscussionSidebar')).toBe('false');
+    } else {
+      expect(localStorage.getItem('showDiscussionSidebar')).toBe('true');
+    }
```
Comment (sundasnoreen12) [inline src/courseware/course/Course.test.jsx]: expect(localStorage.getItem('showDiscussionSidebar')).toBe(result);

File: src/courseware/course/Course.test.jsx [hunk 7]
```diff
@@ -186,6 +218,31 @@ describe('Course', () => {
     expect(screen.getByText(Object.values(models.sequences)[0].title)).toBeInTheDocument();
   });
 
+  [true, false, null].forEach(async (value) => (it(`discussion sidebar is ${value === false ? 'hidden' : 'shown'} when localstorage value is ${value}`, async () => {
```
Comment (sundasnoreen12) [inline src/courseware/course/Course.test.jsx]: use object to pass the hasClass attribute to avoid if else condition

[{value: true, hasClass: null}, {value: false, hasClass: 'd-none'},{value: null, hasClass: null}].forEach(async ({value, hasClass})

File: src/courseware/course/Course.test.jsx [hunk 8]
```diff
@@ -186,6 +218,31 @@ describe('Course', () => {
...
+    await setupDiscussionSidebar(value);
+    const elementTestId = 'sidebar-DISCUSSIONS';
+    await waitFor(() => screen.queryByTestId(elementTestId));
+    const element = screen.queryByTestId(elementTestId);
+    if (value === false) {
+      expect(element).toHaveClass('d-none');
+    } else {
+      expect(element).not.toHaveClass('d-none');
+    }
```
Comment (sundasnoreen12) [inline src/courseware/course/Course.test.jsx]: we can use more precise code like this i think try to reuse it.

const expectation = hasClass ? 'have' : 'not have';
expect(element)[expectation].toHaveClass(hasClass);

**Review by sundasnoreen12 — CHANGES_REQUESTED** [review]

File: src/courseware/course/Course.test.jsx (hunk 1 above)
Comment (muhammadadeeltajamul) [reply src/courseware/course/Course.test.jsx:58]: In learning MFE, course is being created in the factory and it is not defined as a const. It is better to extract course id rather than using a static courseId

File: src/courseware/course/Course.test.jsx (hunk 8 above)
Comment (muhammadadeeltajamul) [reply src/courseware/course/Course.test.jsx]: Element has multiple classes... We need to seperate it using if other wise test case will always fail when it will try to check class null

File: src/courseware/course/sidebar/SidebarContextProvider.jsx [hunk 9]
```diff
@@ -19,9 +19,14 @@ const SidebarProvider = ({
   const shouldDisplayFullScreen = useWindowSize().width < breakpoints.large.minWidth;
   const shouldDisplaySidebarOpen = useWindowSize().width > breakpoints.medium.minWidth;
   const showNotificationsOnLoad = getSessionStorage(`notificationTrayStatus.${courseId}`) !== 'closed';
-  const initialSidebar = (verifiedMode && shouldDisplaySidebarOpen && showNotificationsOnLoad)
+  const discussionStorageEmpty = localStorage.getItem('showDiscussionSidebar') === null;
+  const showDiscussionSidebar = discussionStorageEmpty ? true : localStorage.getItem('showDiscussionSidebar') === 'true';
```
Comment (awais-ansari) [inline src/courseware/course/sidebar/SidebarContextProvider.jsx]: Refactor the code. If can be written.
const showDiscussionSidebar = localStorage.getItem('showDiscussionSidebar') !== 'false';

File: src/courseware/course/sidebar/SidebarContextProvider.jsx [hunk 10]
```diff
@@ -41,6 +46,11 @@ const SidebarProvider = ({
 
   const toggleSidebar = useCallback((sidebarId) => {
     // Switch to new sidebar or hide the current sidebar
+    if (currentSidebar === SIDEBARS.DISCUSSIONS.ID) {
+      localStorage.setItem('showDiscussionSidebar', false);
+    } else if (sidebarId === SIDEBARS.DISCUSSIONS.ID) {
+      localStorage.setItem('showDiscussionSidebar', true);
+    }
     setCurrentSidebar(sidebarId === currentSidebar ? null : sidebarId);
   }, [currentSidebar]);
```
Comment (awais-ansari) [inline src/courseware/course/sidebar/SidebarContextProvider.jsx:54]: Refactor: 
const toggleSidebar = useCallback((newSidebarId) => {
  const showDiscussionSidebar = newSidebarId === SIDEBARS.DISCUSSIONS.ID;
  localStorage.setItem('showDiscussionSidebar', showDiscussionSidebar);
  setCurrentSidebar(currentSidebar === newSidebarId ? null : newSidebarId);
}, [currentSidebar, setCurrentSidebar]);

File: src/courseware/course/Course.test.jsx (hunk 2 above)
Comment (awais-ansari) [inline src/courseware/course/Course.test.jsx]: refactor:
const { courseware: { courseId } } = state;

**Review by sundasnoreen12 — APPROVED** [review]
PR LGTM

File: src/courseware/course/Course.test.jsx [hunk 11]
```diff
@@ -43,6 +49,29 @@ describe('Course', () => {
...
+    const { courseId } = courseware;
+    const axiosMock = new MockAdapter(getAuthenticatedHttpClient());
+    axiosMock.onGet(`${getConfig().LMS_BASE_URL}/api/discussion/v1/courses/${courseId}`).reply(200, { provider: 'openedx' });
+    const topicsResponse = buildTopicsFromUnits(state.models.units);
+    axiosMock.onGet(`${getConfig().LMS_BASE_URL}/api/discussion/v2/course_topics/${courseId}`)
+      .reply(200, topicsResponse);
+
+    await executeThunk(thunks.getCourseDiscussionTopics(courseId), testStore.dispatch);
+    // eslint-disable-next-line prefer-destructuring
+    mockData.unitId = Object.keys(state.models.units)[0];
```
Comment (awais-ansari) [inline src/courseware/course/Course.test.jsx]: const [firstUnitId] = Object.keys(state.models.units);
mockData.unitId = firstUnitId;

This will remove the eslint error

File: src/courseware/course/Course.test.jsx [hunk 12]
```diff
@@ -43,6 +49,29 @@ describe('Course', () => {
...
+    // eslint-disable-next-line prefer-destructuring
+    mockData.sequenceId = Object.keys(state.models.sequences)[0];
```
Comment (awais-ansari) [inline src/courseware/course/Course.test.jsx]: const [firstSequenceId] = Object.keys(state.models.sequences);
mockData.sequenceId = firstSequenceId;

This will remove the eslint error

File: src/courseware/course/Course.test.jsx [hunk 13]
```diff
@@ -186,6 +218,32 @@ describe('Course', () => {
     expect(screen.getByText(Object.values(models.sequences)[0].title)).toBeInTheDocument();
   });
 
+  [
+    { value: true, visible: true },
+    { value: false, visible: false },
+    { value: null, visible: true },
+  ].forEach(async (obj) => (it(`discussion sidebar is ${obj.visible ? 'shown' : 'hidden'} when localstorage value is ${obj.value}`, async () => {
```
Comment (awais-ansari) [inline src/courseware/course/Course.test.jsx]: line length is greater than 120 characters.

Comment (awais-ansari) [reply src/courseware/course/Course.test.jsx]: async ({value, visible}) => {}

directly destructure the object

File: src/courseware/course/Course.test.jsx [hunk 14]
```diff
@@ -186,6 +218,32 @@ describe('Course', () => {
...
+      expect(element).not.toHaveClass('d-none');
+    } else {
+      expect(element).toHaveClass('d-none');
+    }
+  })));
+
+  [
+    { value: true, result: 'false' },
+    { value: false, result: 'true' },
+  ].forEach(async (obj) => (it(`Discussion sidebar storage value is ${!obj.value} when sidebar is ${obj.value ? 'closed' : 'open'}`, async () => {
```
Comment (awais-ansari) [inline src/courseware/course/Course.test.jsx]: line length is greater than 120 characters.
async ({value, visible}) => {}
directly destructure the object

**Review by awais-ansari — CHANGES_REQUESTED** [review]

File: src/courseware/course/sidebar/SidebarContextProvider.jsx (hunk 10 above)
Comment (muhammadadeeltajamul) [reply src/courseware/course/sidebar/SidebarContextProvider.jsx:54]: This code is logically different

**Review by awais-ansari — APPROVED** [review]
